### ISSUE TITLE
Warns about downgrading when upgrading from prerelease

### ISSUE THREAD

**Issue by kraai** [git-for-windows/git#1061]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

Here's the version before the upgrade:

```
$ git --version --build-options
git version 2.11.1.windows.prerelease.2
built from commit: 86d49ccac7afb39e03ba2632803c5ab67d4bb792
sizeof-long: 4
machine: x86_64
```

Here is the version after the upgrade:

```
$ git --version --build-options
git version 2.11.1.windows.1
built from commit: 1c1842bcba45569a84112ec64f72b08eb2d57c68
sizeof-long: 4
machine: x86_64
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 6.1.7601]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
$ cat /etc/install-options.txt
Path Option: Cmd
Plink Path: C:\Program Files (x86)\PuTTY\plink.exe
SSH Option: Plink
CRLF Option: CRLFCommitAsIs
Bash Terminal Option: MinTTY
Performance Tweaks FSCache: Enabled
Use Credential Manager: Enabled
Enable Symlinks: Disabled
Enable Builtin Difftool: Disabled
```

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Bash

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
Git-2.11.1-64-bit.exe
```
 - What did you expect to occur after running these commands?

I expected the installer to start.

 - What actually happened instead?

A dialog containing the following text was displayed:

```
Git for Windows 2.11.1.2 is currently installed.
Do you really want to downgrade to Git for Windows 2.11.1?
```

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

This problem is not occurring with a specific repository.

**Comment by PhilipOakley** [comment]
This is to be expected (unfortunately).

The ordering for the 'pre-releases' looks out of sequence. In fact the pre-release is of the next 'bug' fix of the 2.11.1 series, so it goes 2.11.1, 2.11.1.1, (pre-release) 2.11.1.2, then (if formally released) 2.11.1.2,

Part of the problem is the github releases page's sort order. The one listed at the top of the page isn't the actual (calendar) 'latest', rather it's the last formal release.

**Comment by kraai** [comment]
That doesn't seem right to me.  https://github.com/git-for-windows/git/releases says that the prerelease was released 25 days ago and is 787 commits behind master, whereas the release was released 12 days ago and is only 2 commits behind master.

**Comment by PhilipOakley** [comment]
> That doesn't seem right to me. https://github.com/git-for-windows/git/releases says that the prerelease was released 25 days ago and is 787 commits behind master, whereas the release was released 12 days ago and is only 2 commits behind master.

I stand corrected. My apologies. 

I think the large number of commits behind is also a feature of the GfW rebase process which occurs on top of the Git upstream.

**Comment by dscho** [comment]
The code to determine whether we are downgrading is here: https://github.com/git-for-windows/build-extra/blob/master/installer/install.iss#L560-L603

@kraai maybe you can give it a try to teach that function to understand prereleases?

**Comment by kraai** [comment]
The installed version passed to that function appears to be what's in the registry.  I don't think that is sufficient to decide.  Here's what stored in the registry for some recent versions:

Release Version | Registry Version
------- | ----------------
2.11.0.3 | 2.11.0.3
2.11.1 prerelease 2 | 2.11.1.2
2.11.1 | 2.11.1

If there were a 2.11.1.2 release, I presume the version stored in the registry would be 2.11.1.2, making it indistinguishable from the prerelease.

Should we change the version that's stored in the registry?  Should we use another source for the installed version number (e.g., the mingw-w64-x86_64-git line in C:\Program Files\Git\etc\package-versions.txt)?

**Comment by fourpastmidnight** [comment]
The "pre-release" versions are something that will need to be made to make work better, as the way they are currently versioned will definitely be an issue if/when the MSI installer begins to be another distribution method. For MSI, all four parts of the version number describe a particular version. No other metadata is used to determine the version, unlike for example, SemVer. I've been giving this some thought, but haven't come up with anything concrete.

**Comment by dscho** [comment]
Hrm. @PhilipOakley IIRC it was your suggestion that made me change the way prereleases are versioned.

@PhilipOakley @fourpastmidnight @kraai any suggestions how to name/version those non-official intermediate installers?

**Comment by fourpastmidnight** [comment]
Nothing comes to mind yet. And while working on the MSI last night, I came across some other info about versioning an MSI. Even though all four parts of the version number are used to version an installer, for upgrade purposes, the versions 1.2.3.4 and 1.2.3.5 appear to Windows installer to be version 1.2.3. This is because Windows treats the fourth part of the version number as a "patch" version. Patches are "typically" installed with *.msp files. However, with the component rules being difficult to track, especially with the number of components in Git for Windows, I wouldn't recommend using patches. Using them also adds considerable complexity.

Let me think on it a bit today and get back to you. Any discussion here about the MSI may not have a bearing on the other installers due to MSI's complex versioning requirements which we'll need to discuss at some point in a separate conversation.

**Comment by PhilipOakley** [comment]
> Hrm. @PhilipOakley IIRC it was your suggestion that made me change the way 
> prereleases are versioned.
> @PhilipOakley @fourpastmidnight @kraai any suggestions how to name/version 
> those non-official intermediate installers?

wriggles.. Yes maybe it was me. IIRC it was the difficulty of telling these 
releases from the formal & maintenance releases with the various style of 
semantic versioning (i.e. the nice human incremental numbers). I think at 
the time we were seeing if a 'release' was ready for the big time, and it 
still gave out an old version number, or something like that.

Maybe just add a g<number> instead, same as the main git version string.

It clearly marks them as being different.
They can supercede the previous version, just by having the addendum value.
They don't need to be in strict sequence,
Maybe we should steal it directly from the 'git --version', as the change 
could be elsewhere.

How about using the first five chars of the (pre-)release sha (though the 
https://github.com/git-for-windows/git/wiki/Release-Hashes doesn't say how 
to actually create or check them ;-)


A side question is: how is the order on the Gibhub release page determined?

**Comment by dscho** [comment]
> A side question is: how is the order on the Gibhub release page determined?

No idea.

> Maybe just add a g<number> instead, same as the main git version string.

I find that a little too obscure.

My current thinking is to get rid of the "windows-prerelease" infix and instead add a suffix with incremental numbers on top of the current official version.

At the moment, that would mean that the next preview would be named v2.11.1.windows.1.<something>.1, where I have no idea what word would describe best that this is
1. a non-official, intermediate release,
2. based on v2.11.1.windows.1 but *not* identical to it,
3. *not* a preview for v2.11.1.windows.1, but for whatever comes next

Any ideas what word would do the job?

**Comment by kraai** [comment]
Debian packages use ~ for this situation.  The full documentation can be found at https://www.debian.org/doc/debian-policy/ch-controlfields.html#s-f-Version.  In short, X~Y is sorted before X, so the prereleases could be labelled 2.11.1~prerelease.1, 2.11.1~prerelease.2, etc. and be treated as less than 2.11.1 or 2.11.1.1.

**Comment by PhilipOakley** [comment]
> My current thinking is to get rid of the "windows-prerelease" infix 

I think I'd agree. If we shift the mind set from them being 'pre' and 'post' releases (the former was the one that started the confusion), to having maybe 'bug-fix' (or similar) releases that are increments on the last formal release (as opposed to the _next_ release), then it probably all works.

Adding the `~<num>` (postfix) as @kraai suggest maybe the way out (not had chance to read the doc properly, but the indication looks ok).

**Comment by dscho** [comment]
> X~Y is sorted before X

@kraai my point was: I *tried* that. I *tried* to come up with a strategy to name these prereleases as `<version>-prerelease.<number>` where `<version>` is the upcoming version.

The *problem* is: I do not know what version is next. Up until last Monday, it looked as if v2.11.1(2) was next. However, since I worked on including a couple more bug fixes (and am still on it), it looks more and more as if v2.12.0 (due [tomorrow](http://tinyurl.com/gitCal)) is the next version.

That (not knowing what the next version will be) is the actual root cause for the issue you reported here.

And IMO "update", "bugfix" or even "patch" conveys the message that this is an officially supported version that everybody should install. But that is not what I want. I want to have these semi-regular, intermediate versions that are solely for testing so that I can point bug reporters to an installer they need to test to verify that their problem really has been fixed.

Originally, my idea with the Git for Windows SDK was to ask contributors to build installers from the `master` branch (or from any given Pull Request), but in the meantime I've come to think that this would pose an undue burden on contributors. It's not like we are the Linux kernel project where [we could deliberately waste contributors' time](https://www.youtube.com/watch?v=vyenmLqJQjs&list=PL0lo9MOBetEEm8-hYjdi_A5P2K-eddW47).

If only there were a version format that would transport the message: "This version has advanced further from `<version>` and is here for testing, it already contains changes that will be in the next version, whichever that will be"

**Comment by fourpastmidnight** [comment]
What abou Git-vNext? I mean, what you're talking about is a version of GfW that's not really a released version, but the name still connotes that it comes after the most recent RTW release. And another question, should people be allowed to "upgrade" to a pre-release? Upgrading from pre-release to a released version is one thing; but "upgrading" from a released version to essentially a non-released version is something else entirely---and I propose probably shouldn't be allowed.

**Comment by PhilipOakley** [comment]
> If only there were a version format that would transport the message: "This version has advanced further from `<version>` and is here for testing, it already contains changes that will be in the next version, whichever that will be"

I think there is a logical problem in this full conditional. It is the conflic between "here for testing" and "*will* be in the next version".

Testing implies the possiblity of failure, so it is only a "*may* be included", and that there may be another random version, and another, etc. until a fix is found which could then be included  in the next release. (which is more a statement of how it is, rather than how we'd like it to be ;-)

The other point you make that can be re-inforced is the "This version has advanced further from `<version>` ", and that it should be the basis for adding any suffixes. The 'pre-release` issue is that we fell into the trap of predicting that we had the next release without having its (unkown) number baked into the version string.

One alternate naming thought I had was that, while avoiding "pre-release" and "post-release" which carry too much baggage, we could use "test-release" or similar, as it makes clear what it's purpose is, and it is easy to generate a simple serial number (n.th test-release) as long as we make it plain that they may not be incremental. - e.g. A particular test-release may have just one test feature within it!

In summary, current version, plus test release number.. ?

**Comment by fourpastmidnight** [comment]
Those are great observations. Again though, we should discourage "upgrading to" test releases from a RTW release. Likewise, it _may be possible_ to upgrade from a test release to a future RTW release, but I wonder if out of an abundance of caution we discourage that, too. Discouraging the former with the latter may make the installer logic easier to implement.

**Comment by dscho** [comment]
@fourpastmidnight I actually want to encourage testing, so I am reluctant to disallow switching between official releases and those snapshot installers.

> Testing implies the possiblity of failure,

@PhilipOakley very true!

> so it is only a "may be included",

The thing is: these installers are built from `master`. So the changes *will* be included. Even if the test will require a revert...

I am just thinking of `-interim`. That would make the tag `v2.11.1.windows.1-interim-1`, and the version `v2.11.1-interim-1`. I'll let that sink in for a while, maybe I still like it tomorrow.

**Comment by dscho** [comment]
I settled on a completely different system: there will be continuous snapshots at http://wingit.blob.core.windows.net/files/index.html.

I now have a working VSTS Build setup that was originally only intended to serve the Windows-side of the Continuous Testing of https://github.com/git/git because Travis does not support Windows (and my repeated attempts over one year to get to speak to anybody there simply went unanswered).

That system now works reasonably well, and I just worked on building snapshots whenever `master` advances and publishing them to http://wingit.blob.core.windows.net/files/index.html. This work took an insane amount of effort, forcing me to ignore the Git mailing list a lot more than I am comfortable with, but hey, we have Continuous Snapshots now.

The version numbers used in those snapshots are based on the latest release so that the "do you want to downgrade?" logic should not apply when switching from a snapshot to the latest release and vice versa.
